quark-test.c

```diff
@@ -436,6 +436,103 @@ t_fork_exec_exit(const struct test *t, struct quark_queue_attr *qa)
 	return (0);
 }
 
+static int
+t_process_iterator(const struct test *t, struct quark_queue_attr *qa)
+{
+	struct quark_queue		 qq;
+	struct quark_process_iter	 qi;
+	const struct quark_process	*qp;
+	int				 n_children = 3;
+	pid_t				 children[n_children];
+	struct args			*args;
+	size_t				 expected_len;
+	int				 i;
+	int				 j;
+	char				 cwd[PATH_MAX];
+
+	if (quark_queue_open(&qq, qa) != 0)
+		err(1, "quark_queue_open");
+
+	quark_process_iter_init(&qi, &qq);
+
+	for (i = 0; i < n_children; i++) {
+		children[i] = fork_exec_nop();
+		drain_for_pid(&qq, children[i]);
+	}
+
+	for (i = 0; i < n_children; i++) {
+		do {
+			qp = quark_process_iter_next(&qi);
+		} while (qp != NULL && (pid_t)qp->pid != children[i]);
```

Comment:
The outer loop should be the iterator, and then look for the children.
This is only working now because of how the internal tree is sorted, if that changes, you would be skipping elements.